### ISSUE TITLE
keyPressed broken in processing 3.3.7

### ISSUE THREAD

**Issue by clankill3r** [processing/processing#5441]
```java
void draw() {
}

void keyPressed() {
  println(key); 
}
```

In versions prior to processing 3.3.7 `void keyPressed` gets triggered once a key is pressed.
In processing 3.3.7 the function is called every frame if a key is pressed.
This breaks basically 50% of the sketches I have.

(only tested on Mac OS X 10.12.6)

**Comment by benfry** [comment]
Which renderer are you using? (I don't think any code for `keyPressed()` has changed…)

**Comment by clankill3r** [comment]
This happens with the default / java2d.
I tested it on 2 macs.

**Comment by benfry** [comment]
Hm, I tried with this: 

```processing
void draw() {
  println("new frame");
}

void keyPressed() {
  println(key); 
}

void keyReleased() {
  println("released");
}
```

And got several "new frame" messages between each updated key.

**Comment by gohai** [comment]
I see the same, but with 3.3.6 I only get one invocation of `keyPressed` whereas with 3.3.7 I see multiple (interrupted by new frames).

Could it be that this a change to the VM trying to work around https://github.com/processing/processing/wiki/Troubleshooting#key-repeat-on-macos-sierra ?

**Comment by gohai** [comment]
I found [JDK bug 8177414](https://bugs.openjdk.java.net/browse/JDK-8177414) whose fix was committed September 1 2017, so this could be it.

**Comment by benfry** [comment]
Hm, still can't reproduce. You're using 8u162 / the JDK included with the download, right? 

And out of curiosity, do you have Apple's Java 6 installed? (There have been JVM bugs that only show up when 6 is *not* installed, like the recent menubar problem on High Sierra.)

**Comment by gohai** [comment]
Yes, this happens for me with the 3.3.7 download. (It also happens when I do `ant run` on the current tree).

I see these JVMs installed: `1.6.0.jdk`, `jdk1.8.0_152.jdk`, `jdk1.8.0_162.jdk`

Running `10.13.3`. `defaults read -g ApplePressAndHoldEnabled` gives me back `1`.

**Comment by benfry** [comment]
What happens if you set it to zero or delete it? (I get `The domain/default pair of (kCFPreferencesAnyApplication, ApplePressAndHoldEnabled) does not exist`)

**Comment by gohai** [comment]
Same thing. (I did restart just to be sure.)

To clarify: on `3.3.6`, when I press and hold a letter-key, I only see one invocation of `keyPressed` while I do so. (I see multiple calls to `keyPressed` when I press and hold Space, or punctuation.) On `3.3.7`, I get a steady stream of `keyPressed` for pressing and holding a letter-key.

**Comment by gohai** [comment]
```
void setup() {
  frameRate(1);
}

void draw() {
  if (keyPressed) {
    println("Draw w/ " + key);
  } else {
    println("Draw w/o key");
  }
}

void keyPressed() {
  println("Pressed w/ " + key);
}

void keyReleased() {
  println("Released w/ " + key);
}
```

This is me holding `g` for a second on 3.3.7:

```
Draw w/o key
Pressed w/ g
Draw w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Pressed w/ g
Draw w/ g
Pressed w/ g
Pressed w/ g
Released w/ g
Draw w/o key
Draw w/o key
Draw w/o key
```

(So it seems at least the `keyReleased` is happening only once.)

**Comment by benfry** [comment]
But that looks correct to me… Due to key repeat, pressing and holding will cause `keyPressed` to fire multiple times, as long as you have the key down. What am I missing?

**Comment by gohai** [comment]
The behavior on 3.3.6 with the same input:

```
Draw w/o key
Pressed w/ g
Draw w/ g
Draw w/ g
Draw w/ g
Draw w/ g
Released w/ g
Draw w/o key
Draw w/o key
Draw w/o key
```

So at least on 3.3.6 the `keyPressed` and `keyReleased` were symmetrical, and the Reference also says: "The keyPressed() function is called _once_ every time a key is pressed."

**Comment by benfry** [comment]
Or if you use `hint(DISABLE_KEY_REPEAT)`? (Though I think that only works for OpenGL)

**Comment by kfrajer** [comment]
@clankill3r  and @gohai  My 2c.... would it possible to revert to 3.3.6 and test if the `keyPressed()` has the behavior of only one call(instead of multiple calls observed in 3.3.7? This will test for sure it is Processing and not a recent update from Java. I am on Windows so I cannot provide a test for your case. However, I do want to comment that in Windows:

keyPressed() will only print once.

Kf

**Comment by kfrajer** [comment]
Ok, so I did a check in both Processing 3.3.6 and 3.3.7 and `keyPressed` returns multiple entries (Not a single one as I reported before). This is in Windows 10 x64. To add to this, I just recently updated my java:

> C:\Users\C\Desktop>java -version
> java version "1.8.0_161"
> Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
> Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)

Online java version check:  https://www.java.com/en/download/installed.jsp

Kf

**Comment by JakubValtar** [comment]
On Windows, holding a key down always behaved like in the text editor:
- keyPressed & keyTyped, long pause
- repeated keyPressed & keyTyped with short pauses
- keyReleased

**Comment by gohai** [comment]
If the behavior that @JakubValtar describes is the expected then I think we can close this bug, as this is exactly what is now happening on macOS also.

@benfry I haven't tested by I strongly believe that it is solely the JVM that changed behavior on macOS. On the plus side: this finally fixes [Key repeat on macOS Sierra](https://github.com/processing/processing/wiki/Troubleshooting#key-repeat-on-macos-sierra), so we can get rid of this awkward workaround.

**Comment by benfry** [comment]
I guess most of the default behaviors are weird and the "correct" behavior (multiple `mousePressed()`, only one `mouseReleased()`) aren't what users expect. 

We have some control over it in OpenGL, and we could potentially hack it with Java2D and JavaFX… So it might be worth considering whether we do: 

* `keyPressed()` and `keyReleased()` only for physical down/up events
* `keyTyped()` is subject to key repeat; anyone implementing something that needs repeat (i.e. a text field) should be using `keyTyped()` anyway

In the past we couldn't do this because it was a lot of extra effort to paper over slight differences by renderer and OS, but with @JakubValtar's additional work on tracking key presses in there, we might not be far from this kind fo implementation. (And we could get rid of the `KEY_REPEAT` hints.)

**Comment by clankill3r** [comment]
I agree with:

* keyPressed() and keyReleased() only for physical down/up events

For me it has always been that way, and I think most people expect it this
way.


2018-03-20 16:34 GMT+01:00 Ben Fry <notifications@github.com>:

**Comment by clankill3r** [comment]
Mac OS X 10.12.6
Processing 3.4

I did a test to see when `keyPressed` and `keyTyped` do repeat.
There is **no** difference in behaviour between  `keyPressed` and `keyTyped`.
Those are the results, a ✓ means it repeats.

|                | VSCODE | PDE |
|----------|-----|-----|
| FX2D     | ✓      | ✓   |
| JAVA2D  | ✘      | ✓   |
| P2D        | ✘      | ✘   |

**Comment by clankill3r** [comment]
I did some more testing.
Apple has a accent menu when holding a key:

<img width="213" alt="screen shot 2018-08-19 at 12 42 33" src="https://user-images.githubusercontent.com/738650/44308076-fd0e2600-a3ae-11e8-8ae1-840e382c6332.png">

I tested, and it only would never trigger a keyPressed again when holding one of those:
`e, y, u, I, o, a, s, l, z, c, n`. (Those are the ones with a accent menu).
I also noticed that if you hold a key that has no accents, the accent menu shows up for like 1 frame.

I disabled the accent menu using:

`defaults write -g ApplePressAndHoldEnabled -bool false`

The behaviour for the PDE is now the same as in VSCode.
Also there is still no difference between keyPressed and keyTyped.

For the rest, when holding a key:

FX2D - repeats
JAVA2D - repeats
P2D - no repeat

**Comment by JakubValtar** [comment]
I think #3652 explains what you see.

OpenGL does not have keyTyped, so we try to sythesize it from keyPressed. Also we block auto-repeat in OpenGL, since the platform marks the repeated presses for us and we just filter them. We don't have that implemented for JAVA2D and FX2D yet, since we don't get the repeat info from the platform and we would have to do it manually.

Here is a wiki entry for ApplePressAndHoldEnabled https://github.com/processing/processing/wiki/Troubleshooting#key-repeat-on-macos-sierra

Do you propose some changes which would make it less confusing, @clankill3r?
